File: cpp/capstone/src/command.cpp
// command.cpp — РЕАЛИЗАЦИЯ разбора команд. Сейчас СТАБ: компилируется, но всегда отвечает "ERR",
// поэтому tests/test_command.cpp стартует КРАСНЫМ.
//
// Подсказки по реализации (детали — спрашивай меня):
//   - Разбей line на токены по пробелам (std::istringstream + оператор >> — модуль 08).
//   - Имя команды приведи к верхнему регистру, затем switch/if по нему.
//   - Аргументы-числа парси аккуратно: std::stol в try/catch (модуль 09) → при ошибке "ERR ...".
//   - Список/хеш-ответы склеивай через '\n'; пустые — "(empty)".
//   - НИКОГДА не давай исключению вылететь наружу: оберни тело в try/catch и верни "ERR ...".

#include "command.hpp"

#include <algorithm>
#include <cctype>
#include <sstream>
#include <string>
#include <vector>

namespace minidb {

namespace {

// Разбить строку на токены по пробельным символам (istringstream + operator>>).
std::vector<std::string> tokenize(const std::string& line) {
    std::vector<std::string> tokens;
    std::istringstream iss(line);
    std::string tok;
    while (iss >> tok) tokens.push_back(tok);
    return tokens;
}

// Имя команды к ВЕРХНЕМУ регистру: set и SET — одна команда.
std::string upper(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(),
                   [](unsigned char c) { return static_cast<char>(std::toupper(c)); });
    return s;
}

// Склеить вектор строк через '\n'; пустой → "(empty)". Используется для LRANGE/HKEYS/KEYS.
std::string join_lines(const std::vector<std::string>& items) {
    if (items.empty()) return "(empty)";
    std::string out;
    for (std::size_t i = 0; i < items.size(); ++i) {
        if (i) out += '\n';
        out += items[i];
    }
    return out;
}

// Распарсить целое из токена строго (вся строка — число). Бросает при мусоре.
long parse_long(const std::string& s) {
    std::size_t pos = 0;
    long v = std::stol(s, &pos);  // бросит std::invalid_argument на нечисле
    if (pos != s.size()) throw std::invalid_argument("trailing chars");
    return v;
}

}  // namespace

// execute ВСЕГДА возвращает строку — любое исключение ловим и превращаем в "ERR ...".
// Это держит REPL и тесты простыми: слой команд не «протекает» исключениями наружу.
std::string execute(Database& db, const std::string& line) {
    try {
        auto tokens = tokenize(line);
        if (tokens.empty()) return "ERR empty command";

        const std::string cmd = upper(tokens[0]);
        const std::size_t argc = tokens.size() - 1;  // число аргументов без имени команды

        // Маленький помощник: число с префиксом ':' (Redis-стиль для целочисленных ответов).
        auto integer = [](long n) { return ":" + std::to_string(n); };

        // ------------------------------ strings ------------------------------
        if (cmd == "SET") {
            if (argc != 2) return "ERR wrong number of arguments";
            db.set(tokens[1], tokens[2]);
            return "OK";
        }
        if (cmd == "GET") {
            if (argc != 1) return "ERR wrong number of arguments";
            auto v = db.get(tokens[1]);
            return v ? *v : "(nil)";
        }

        // ------------------------------- lists -------------------------------
        if (cmd == "LPUSH") {
            if (argc != 2) return "ERR wrong number of arguments";
            return integer(static_cast<long>(db.lpush(tokens[1], tokens[2])));
        }
        if (cmd == "LRANGE") {
            if (argc != 3) return "ERR wrong number of arguments";
            long start = parse_long(tokens[2]);
            long stop = parse_long(tokens[3]);
            return join_lines(db.lrange(tokens[1], start, stop));
        }
        if (cmd == "LLEN") {
            if (argc != 1) return "ERR wrong number of arguments";
            return integer(static_cast<long>(db.llen(tokens[1])));
        }

        // ------------------------------- hashes ------------------------------
        if (cmd == "HSET") {
            if (argc != 3) return "ERR wrong number of arguments";
            return integer(db.hset(tokens[1], tokens[2], tokens[3]) ? 1 : 0);
        }
        if (cmd == "HGET") {
            if (argc != 2) return "ERR wrong number of arguments";
            auto v = db.hget(tokens[1], tokens[2]);
            return v ? *v : "(nil)";
        }
        if (cmd == "HKEYS") {
            if (argc != 1) return "ERR wrong number of arguments";
            return join_lines(db.hkeys(tokens[1]));
        }

        // ------------------------------ generic ------------------------------
        if (cmd == "DEL") {
            if (argc != 1) return "ERR wrong number of arguments";
            return integer(db.del(tokens[1]) ? 1 : 0);
        }
        if (cmd == "TYPE") {
            if (argc != 1) return "ERR wrong number of arguments";
            auto t = db.type(tokens[1]);
            if (!t) return "none";
            switch (*t) {
                case Type::String: return "string";
                case Type::List:   return "list";
                case Type::Hash:   return "hash";
            }
            return "none";
        }
        if (cmd == "KEYS") {
            if (argc != 0) return "ERR wrong number of arguments";
            return join_lines(db.keys());
        }
        if (cmd == "COUNT") {
            if (argc != 0) return "ERR wrong number of arguments";
            return integer(static_cast<long>(db.size()));
        }

        // ------------------------------- expiry ------------------------------
        if (cmd == "EXPIRE") {
            if (argc != 2) return "ERR wrong number of arguments";
            long secs = parse_long(tokens[2]);  // нечисло → catch → "ERR value is not an integer"
            return integer(db.expire(tokens[1], static_cast<std::int64_t>(secs)) ? 1 : 0);
        }
        if (cmd == "TTL") {
            if (argc != 1) return "ERR wrong number of arguments";
            return integer(static_cast<long>(db.ttl(tokens[1])));
        }
        if (cmd == "PERSIST") {
            if (argc != 1) return "ERR wrong number of arguments";
            return integer(db.persist(tokens[1]) ? 1 : 0);
        }

        // ---------------------------- persistence ----------------------------
        if (cmd == "SAVE") {
            if (argc != 1) return "ERR wrong number of arguments";
            return db.save(tokens[1]) ? "OK" : "ERR save failed";
        }
        if (cmd == "LOAD") {
            if (argc != 1) return "ERR wrong number of arguments";
            return db.load(tokens[1]) ? "OK" : "ERR load failed";
        }

        return "ERR unknown command";
    } catch (const std::invalid_argument&) {
        // Сюда падает parse_long на нечисловом аргументе.
        return "ERR value is not an integer";
    } catch (const std::out_of_range&) {
        // std::stol при переполнении long.
        return "ERR value is not an integer";
    } catch (const std::exception&) {
        // Любая иная ошибка (например, бросающий аксессор Value) не должна вылетать наружу.
        return "ERR internal error";
    }
}

}  // namespace minidb

File: rust/capstone/src/command.rs
//! Модуль `command` — текстовый протокол: разбор строки и исполнение команды.
//!
//! [`execute`] — это «мозг» REPL и сетевого слоя: на входе одна строка вроде
//! `"SET name alice"`, на выходе строковый ответ в Redis-подобном стиле (`"OK"`,
//! `"(nil)"`, `"(integer) 3"`, …). Здесь нет состояния — только разбор и вызов методов
//! [`crate::Database`]. Тот же `execute` используют и [`crate::SharedDb::exec`], и
//! бинарник `main.rs`.

use crate::db::Database;

/// Разобрать и исполнить одну команду над хранилищем, вернув текстовый ответ.
///
/// Команда — это слова, разделённые пробелами. Первое слово — имя команды
/// (регистронезависимо: `SET`, `set`, `Set` — одно и то же). Остальные слова —
/// аргументы. Значение для `SET` / `LPUSH` / `HSET` — это **последний** аргумент
/// (для простоты считаем, что значения без пробелов; обработку кавычек можешь добавить
/// как стретч-цель).
///
/// Поддерживаемые команды и их ответы (эталон — приёмочные тесты):
///
/// | Команда                 | Успех                                  | Особые случаи                          |
/// |-------------------------|----------------------------------------|----------------------------------------|
/// | `SET key value`         | `OK`                                   |                                        |
/// | `GET key`               | значение                               | нет ключа/не строка → `(nil)`          |
/// | `DEL key`               | `(integer) 1`                          | ключа не было → `(integer) 0`          |
/// | `TYPE key`              | `string` \| `list` \| `hash`           | нет ключа → `none`                     |
/// | `KEYS`                  | ключи через пробел (сорт.)             | пусто → `(empty)`                      |
/// | `LPUSH key value`       | `(integer) <новая длина>`              | неверный тип → `(error) WRONGTYPE ...` |
/// | `LRANGE key start stop` | элементы через пробел                  | пусто → `(empty)`                      |
/// | `LLEN key`              | `(integer) <длина>`                    |                                        |
/// | `HSET key field value`  | `(integer) 1` (новое) / `(integer) 0`  | неверный тип → `(error) WRONGTYPE ...` |
/// | `HGET key field`        | значение                               | нет → `(nil)`                          |
/// | `HKEYS key`             | поля через пробел (сорт.)              | пусто → `(empty)`                      |
/// | `SAVE path`             | `OK`                                   | ошибка I/O → `(error) <текст>`         |
/// | `LOAD path`             | `OK`                                   | ошибка I/O → `(error) <текст>`         |
///
/// Общие правила ответов:
/// - неизвестная команда → `(error) unknown command '<имя>'`;
/// - нехватка аргументов → `(error) wrong number of arguments for '<имя>'`;
/// - в `LRANGE` нечисловые `start`/`stop` → `(error) value is not an integer`;
/// - для `Err(msg)` от методов базы (неверный тип) ответ — `(error) <msg>`.
///
/// ПОЧЕМУ всё это возвращается строкой, а не печатается тут же: так `execute` остаётся
/// чистой функцией (вход → выход), её легко тестировать и переиспользовать в потоках.
/// Реализуй через разбор `line.split_whitespace()` и `match` по имени команды (модуль 06),
/// прокидывая ошибки методов базы (модуль 07) в формат `(error) ...`.
pub fn execute(db: &mut Database, line: &str) -> String {
    let args: Vec<&str> = line.split_whitespace().collect();
    if args.is_empty() {
        return String::new();
    }
    let cmd = args[0].to_uppercase();
    match cmd.as_str() {
        "SET" => {
            if args.len() < 3 {
                return format!("(error) wrong number of arguments for 'SET'");
            }
            db.set(args[1], args[2].to_string());
            "OK".to_string()
        }
        "GET" => {
            if args.len() < 2 {
                return format!("(error) wrong number of arguments for 'GET'");
            }
            match db.get(args[1]) {
                Some(s) => s,
                None => "(nil)".to_string(),
            }
        }
        "DEL" => {
            if args.len() < 2 {
                return format!("(error) wrong number of arguments for 'DEL'");
            }
            let deleted = db.del(args[1]);
            format!("(integer) {}", if deleted { 1 } else { 0 })
        }
        "TYPE" => {
            if args.len() < 2 {
                return format!("(error) wrong number of arguments for 'TYPE'");
            }
            match db.type_of(args[1]) {
                Some(t) => t.to_string(),
                None => "none".to_string(),
            }
        }
        "KEYS" => {
            let keys = db.keys();
            if keys.is_empty() {
                "(empty)".to_string()
            } else {
                keys.join(" ")
            }
        }
        "LPUSH" => {
            if args.len() < 3 {
                return format!("(error) wrong number of arguments for 'LPUSH'");
            }
            match db.lpush(args[1], args[2].to_string()) {
                Ok(len) => format!("(integer) {len}"),
                Err(msg) => format!("(error) {msg}"),
            }
        }
        "LRANGE" => {
            if args.len() < 4 {
                return format!("(error) wrong number of arguments for 'LRANGE'");
            }
            let start = match args[2].parse::<i64>() {
                Ok(n) => n,
                Err(_) => return "(error) value is not an integer".to_string(),
            };
            let stop = match args[3].parse::<i64>() {
                Ok(n) => n,
                Err(_) => return "(error) value is not an integer".to_string(),
            };
            let elems = db.lrange(args[1], start, stop);
            if elems.is_empty() {
                "(empty)".to_string()
            } else {
                elems.join(" ")
            }
        }
        "LLEN" => {
            if args.len() < 2 {
                return format!("(error) wrong number of arguments for 'LLEN'");
            }
            format!("(integer) {}", db.llen(args[1]))
        }
        "HSET" => {
            if args.len() < 4 {
                return format!("(error) wrong number of arguments for 'HSET'");
            }
            match db.hset(args[1], args[2], args[3].to_string()) {
                Ok(is_new) => format!("(integer) {}", if is_new { 1 } else { 0 }),
                Err(msg) => format!("(error) {msg}"),
            }
        }
        "HGET" => {
            if args.len() < 3 {
                return format!("(error) wrong number of arguments for 'HGET'");
            }
            match db.hget(args[1], args[2]) {
                Some(v) => v,
                None => "(nil)".to_string(),
            }
        }
        "HKEYS" => {
            if args.len() < 2 {
                return format!("(error) wrong number of arguments for 'HKEYS'");
            }
            let keys = db.hkeys(args[1]);
            if keys.is_empty() {
                "(empty)".to_string()
            } else {
                keys.join(" ")
            }
        }
        "SAVE" => {
            if args.len() < 2 {
                return format!("(error) wrong number of arguments for 'SAVE'");
            }
            match db.save(args[1]) {
                Ok(()) => "OK".to_string(),
                Err(e) => format!("(error) {e}"),
            }
        }
        "LOAD" => {
            if args.len() < 2 {
                return format!("(error) wrong number of arguments for 'LOAD'");
            }
            match db.load(args[1]) {
                Ok(()) => "OK".to_string(),
                Err(e) => format!("(error) {e}"),
            }
        }
        _ => {
            format!("(error) unknown command '{}'", cmd)
        }
    }
}
